### PR TITLE
[IMP] product: Add helper methods to compute and set product variant fields

### PR DESCRIPTION
**Pull request by ivantodorovich**
As a bonus:

* Use the `Query` returned by `_search` when searching on `barcode` (this uses a subquery to search, saves a db roundtrip).
* Simplify search by standard_price, should also save 1 query.

Optional to consider: Set `store=False` on `weight` and `volume`, replacing it with a `search='_search_*`

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/product/models/product_template.py [hunk 1]
```diff
@@ -179,70 +179,72 @@ def _compute_currency_id(self):
...
-            elif variant_count == 0:
-                archived_variants = template.with_context(active_test=False).product_variant_ids
-                if len(archived_variants) == 1:
-                    template.barcode = archived_variants.barcode
+        return self._compute_product_variant_field('barcode')
 
     def _search_barcode(self, operator, value):
-        templates = self.with_context(active_test=False).search([('product_variant_ids.barcode', operator, value)])
-        return [('id', 'in', templates.ids)]
+        templates = self.with_context(active_test=False)._search([('product_variant_ids.barcode', operator, value)])
```
Comment (Feyensv) [inline addons/product/models/product_template.py:243]:
```suggestion
        subquery = self.with_context(active_test=False)._search([('product_variant_ids.barcode', operator, value)])
```

File: addons/product/models/product_template.py [hunk 2]
```diff
@@ -179,70 +179,72 @@ def _compute_currency_id(self):
...
     def _set_standard_price(self):
-        for template in self:
-            if len(template.product_variant_ids) == 1:
-                template.product_variant_ids.standard_price = template.standard_price
+        return self._set_product_variant_field('standard_price')
 
     def _search_standard_price(self, operator, value):
-        products = self.env['product.product'].search([('standard_price', operator, value)], limit=None)
-        return [('id', 'in', products.mapped('product_tmpl_id').ids)]
+        return [('product_variant_ids.standard_price', operator, value)]
```
Comment (Feyensv) [inline addons/product/models/product_template.py:219]: Please split this change in a separate commit (with the other `_search` methods changes)

File: addons/product/models/product_template.py [hunk 3]
```diff
@@ -179,70 +179,72 @@ def _compute_currency_id(self):
...
     def _compute_standard_price(self):
-        # Depends on force_company context because standard_price is company_dependent
-        # on the product_product
-        unique_variants = self.filtered(lambda template: len(template.product_variant_ids) == 1)
-        for template in unique_variants:
-            template.standard_price = template.product_variant_ids.standard_price
-        for template in (self - unique_variants):
-            template.standard_price = 0.0
+        # Depends on force_company context because standard_price is company_dependent on the product_product
+        return self._compute_product_variant_field('standard_price')
```
Comment (Feyensv) [inline addons/product/models/product_template.py:213]:
```suggestion
        self._compute_product_variant_field('standard_price')
```

drop all `return` calls please, i prefer to make it clear that the compute and inverse methods do not return anything.

File: addons/product/models/product_template.py [hunk 4]
```diff
@@ -179,70 +179,72 @@ def _compute_currency_id(self):
     def _compute_cost_currency_id(self):
         self.cost_currency_id = self.env.company.currency_id.id
 
+    def _compute_product_variant_field(self, fname, default=False):
```
Comment (Feyensv) [inline addons/product/models/product_template.py:182]:
```suggestion
    def _compute_template_field_from_variant(self, fname, default=False):
```

File: addons/product/models/product_template.py [hunk 5]
```diff
@@ -179,70 +179,72 @@ def _compute_currency_id(self):
...
+        """Computes a field from the template's product.product
```
Comment (Feyensv) [inline addons/product/models/product_template.py:183]:
```suggestion
        """Sets the value of the given field based on the template variant values
```

File: addons/product/models/product_template.py [hunk 6]
```diff
@@ -179,70 +179,72 @@ def _compute_currency_id(self):
...
+
+        Equals to product_variant_ids[fname] if it's a single variant product.
+        Otherwise, sets the value specified in ``default``.
+        It's used to compute fields like barcode, weight, volume..
```
Comment (Feyensv) [inline addons/product/models/product_template.py:187]:
```suggestion
        It's used to compute fields like barcode, weight, volume..
        
        :param str fname: name of the field to compute
            (field name must be identical between product.product & product.template models)
        :param default: default value to set when there are multiple or no variants on the template
        :return: None
```

File: addons/product/models/product_template.py [hunk 7]
```diff
@@ -179,70 +179,72 @@ def _compute_currency_id(self):
...
+    @api.depends('product_variant_ids.standard_price')
     def _compute_standard_price(self):
-        # Depends on force_company context because standard_price is company_dependent
-        # on the product_product
-        unique_variants = self.filtered(lambda template: len(template.product_variant_ids) == 1)
-        for template in unique_variants:
-            template.standard_price = template.product_variant_ids.standard_price
-        for template in (self - unique_variants):
-            template.standard_price = 0.0
+        # Depends on force_company context because standard_price is company_dependent on the product_product
```
Comment (Feyensv) [inline addons/product/models/product_template.py:212]:
```suggestion
        # Depends on force_company context because standard_price is company_dependent
        # on the product_product
```

Keep it on two lines pls, to respect a max line length of 100 characters

**Review by Feyensv — COMMENTED** [review]
Nice improvements :+1: 

Could you split the two changes in two different commits please ?
* the perf diff on the _search methods
* the factorization of the compute & inverse methods

I also gave some linting/docstring/guidelines comments on the fly :).

**Comment by ivantodorovich** [comment]
Thanks for your review and your suggestions @Feyensv !
The changes have been applied ;)

File: addons/product/models/product_template.py [hunk 8]
```diff
@@ -179,70 +179,84 @@ def _compute_currency_id(self):
...
+                template[fname] = default
+
+    def _set_product_variant_field(self, fname):
+        """Set a field to the template's product.product
+
+        Only if it's a single variant product.
+        It's used to set fields like barcode, weight, volume..
+
+        :param str fname: name of the field to compute
+            (field name must be identical between product.product & product.template models)
```
Comment (Feyensv) [inline addons/product/models/product_template.py:212]:
```suggestion
        :param str fname: name of the field whose value should be propagated to the variant.
            (field name must be identical between product.product & product.template models)
```

File: addons/product/models/product_template.py [hunk 9]
```diff
@@ -179,70 +179,84 @@ def _compute_currency_id(self):
...
+                template[fname] = template.product_variant_ids[fname]
+            elif variant_count == 0 and self.env.context.get("active_test", True):
+                # If the product has no active variants, retry without the active_test
+                template_ctx = template.with_context(active_test=False)
+                template_ctx._compute_template_field_from_variant_field(fname, default=default)
+            else:
+                template[fname] = default
+
+    def _set_product_variant_field(self, fname):
+        """Set a field to the template's product.product
```
Comment (Feyensv) [inline addons/product/models/product_template.py:206]:
```suggestion
        """Propagate the value of the given field from the templates to their unique variant.
```

File: addons/product/models/product_template.py [hunk 10]
```diff
@@ -179,70 +179,84 @@ def _compute_currency_id(self):
...
-                if len(archived_variants) == 1:
-                    template.barcode = archived_variants.barcode
+        self._compute_template_field_from_variant_field('barcode')
 
     def _search_barcode(self, operator, value):
-        templates = self.with_context(active_test=False).search([('product_variant_ids.barcode', operator, value)])
-        return [('id', 'in', templates.ids)]
+        subquery = self.with_context(active_test=False)._search(
+            [('product_variant_ids.barcode', operator, value)]
+        )
```
Comment (Feyensv) [inline addons/product/models/product_template.py:255]:
```suggestion
        subquery = self.with_context(active_test=False)._search([
            ('product_variant_ids.barcode', operator, value),
        ])
```
we tend to keep one domain leaf for each line when we indent search calls.
That way, when we add a new leaf, there is only one modified line in the git diff :).

**Review by Feyensv — COMMENTED** [review]
Last linting remarks on some docstrings/code (hopefully).
Thanks for the changes :).

**Comment by ivantodorovich** [comment]
Done ^^

**Review by Feyensv — APPROVED** [review]
@robodoo r+ rebase-ff

**Comment by Feyensv** [comment]
About the optional part and removing the `store=True` on `product.template` weigth and volume, it might be worth investigating to see the performance impact and if those fields are used somewhere in the code.
If it's only used in the views (search view include), i guess storing them is not that useful.

**Comment by ivantodorovich** [comment]
> About the optional part and removing the `store=True` on `product.template` weight and volume, it might be worth investigating to see the performance impact and if those fields are used somewhere in the code. If it's only used in the views (search view include), i guess storing them is not that useful.

AFAICS they're not really used

The impact is not huge, but having them not stored would save a few recomputes+writes when creating/modifing variants

**Comment by Feyensv** [comment]
Well, i suppose it can stay like this for now, i'll try to keep it in mind for the next cleanup/apocalypse in the `product` scope ;).
